### PR TITLE
[Core] Remote Instance Model Loader: A loader used to fetch model weights from a ready vLLM instance.

### PR DESCRIPTION
**Pull request by lirong-lirong**
## Purpose

This PR implements a new model loading mechanism that enables transferring model weights from remote vLLM instances. This feature is particularly useful in distributed deployments where multiple vLLM instances need to share model weights without individually loading them from storage.

Key features of this implementation:
- Adds `RemoteInstanceModelLoader` for loading model weights from GPUs of other vLLM instances
- Supports weight transfer between instances with the same parallelism strategy (include TP and PP)
- Implements custom process groups for secure and efficient weight transfer
- Adds new API endpoints (`/init_weights_send_group_for_remote_instance` and `/send_weights_to_remote_instance`) for coordinating weight transfers
- Extends the model loader framework to support the new "remote_instance" load format
- Includes validation mechanisms to ensure model compatibility between instances
- ​​Leverages high-speed interconnects (NVLink/RDMA) for efficient model weight distribution and loading​

This PR effectively reduces the startup time of vLLM, particularly in large-scale scaling scenarios on the cloud.

## Test Plan

End-to-end tests with multiple vLLM instances to verify weight transfer

Test commands:
```bash
pytest -s -v tests/model_executor/model_loader/test_remote_instance_loader.py
```

## Test Result
[log.txt](https://github.com/user-attachments/files/23098595/log.txt)

### Start a vLLM instance normally as a seed instance to provide the model weight data source

```bash
python3 -m vllm.entrypoints.openai.api_server \
	--model /data0/deepseek-ai/DeepSeek-V3.1 \
	--tensor-parallel-size 8 \
	--host 0.0.0.0 --port 12346 \
	--gpu-memory-utilization 0.95 \
	--max_model_len 1024 \
	--enforce_eager
```

### Launch new vLLM using the Remote Instance Model Loader
```bash
export REMOTE_INSTANCE_IP="10.32.11.46"
export REMOTE_INSTANCE_SERVER_PORT=12346
export REMOTE_INSTANCE_PORTS="[62000,62001,62002,62003,62004,62005,62006,62007]"

python3 -m vllm.entrypoints.openai.api_server \
	--model /data0/deepseek-ai/DeepSeek-V3.1 \
        --load_format remote_instance \
	--tensor-parallel-size 8 \
	--port 12346 \
	--gpu-memory-utilization 0.95 \
	--max_model_len 1024 \
	--enforce_eager
```

The three environment variables above represent the IP address of the remote instance, the service port, and the designated ports for the NCCL process group, respectively. The number of ports should match the number of workers.

## Additional Context

We are continuously working on reducing the startup overhead of inference services. Previously, we implemented other solutions, such as using Mooncake Store to store processed weight tensors and pulling them directly to the GPU via GPUDirect RDMA (GDR), as well as leveraging the Mooncake transfer engine for direct GPU-to-GPU tensor transfer between old and new instances.

Recently, the sglang project introduced a new 'remote instance model loader' feature, which accomplishes the same task using the default NCCL communication backend combined with a broadcast operation. Considering the potential value of maintaining future compatibility with sglang, we have implemented the same approach in vLLM.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by ovidiusm** [comment]
This transfer should reuse vllm's `NixlConnector` instead of creating a new transfer method from scratch with torch collectives.

There are serverless usecases where the NIXL API is preferrable, since it allows, for example, fast swapping in/out of weights to cache on workers that serve several models and are oversubscribed in terms of VRAM, with models used infrequently. There are already some projects working on this.

Reusing NixlConnector will also bring for free all the performance optimizations made for KVCache transfers.

File: vllm/distributed/kv_transfer/kv_connector/v1/weight_transfer_connector.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        client_rank: int = -1,
+        client_id: str = "",
+        group_rank: int = 1,
+        world_size: int = 2,
+    ):
+        assert gpu_id != -1 and client_rank != -1, (
+            "gpu_id and tp_rank must be specified for RemoteInstanceConnector. "
+        )
+
+        self.device_id = torch.device("cuda", gpu_id)
```
Comment (ApsarasX) [inline vllm/distributed/kv_transfer/kv_connector/v1/weight_transfer_connector.py:46]: use current_platform rather than `cuda`

File: vllm/model_executor/model_loader/remote_instance_loader.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        self,
+        model: nn.Module,
+        client: WeightTransferConnector,
+        model_config: ModelConfig,
+    ) -> None:
+        start_build_group_tic = time.time()
+        # To support tp, pp
+        global_rank = _get_rank()
+        success, message = client.build_group(
+            gpu_id=torch.cuda.current_device(),
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader.py:107]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                    _get_instance_service_port(),
+                    _get_instance_ports(),
+                    self.client_id,
+                    sum(1 for v in model.state_dict().values() if v.numel() > 0),
+                ),
+            )
+            t.start()
+
+        try:
+            torch.cuda.empty_cache()
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader.py:141]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+                            group_src=0,
+                            group=client._model_update_group,
+                        )
+
+            end_get_weights_tic = time.time()
+            logger.info(
+                "finish getting all weights from remote instance, time used: %.4fs",
+                end_get_weights_tic - start_get_weights_tic,
+            )
+            torch.cuda.empty_cache()
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader.py:159]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader_utils.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    tp_size = get_tp_group().world_size
+
+    from vllm.distributed.parallel_state import get_pp_group
+
+    pp_rank = get_pp_group().rank_in_group
+    pp_size = get_pp_group().world_size
+    global_rank = pp_rank * tp_size + tp_rank
+
+    ports_list = ports.split(",")
+    gpu_id = torch.cuda.current_device()
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader_utils.py:161]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader_utils.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        from datetime import timedelta
+
+        _weights_send_group[group_name] = init_custom_process_group(
+            backend=backend,
+            init_method=f"tcp://{master_address}:{group_port}",
+            timeout=timedelta(seconds=60),
+            world_size=world_size,
+            rank=group_rank,
+            group_name=group_name,
+            device_id=torch.device("cuda", gpu_id),
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader_utils.py:195]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader_utils.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        if send_group is None:
+            message = (
+                f"Group {group_name} not in _weights_send_group list. "
+                f"Please call `init_weights_send_group_for_remote_instance` first."
+            )
+            logger.error(message)
+            return {"success": False, "message": message}
+
+        logger.info("Send weight in %s", send_group)
+        torch.cuda.empty_cache()
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader_utils.py:290]: ditto

File: vllm/model_executor/model_loader/remote_instance_loader_utils.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        state_dict = model.state_dict()
+        for key, tensor in state_dict.items():
+            if tensor.numel():
+                torch.distributed.broadcast(
+                    tensor,
+                    group_src=0,
+                    group=send_group,
+                )
+        torch.cuda.empty_cache()
```
Comment (ApsarasX) [inline vllm/model_executor/model_loader/remote_instance_loader_utils.py:299]: ditto

**Review by ApsarasX — CHANGES_REQUESTED** [review]
